### PR TITLE
Reply "insufficient storage" on disk full or over-quota

### PR DESCRIPTION
**Pull request by Enrico204**
What is the purpose of this change? What does it change?
--------------------------------------------------------

This pull request will change the current behavior on disk-related errors:
* HTTP 507 "Insufficient storage" is the status on disk full or over-quota
* HTTP 500 "Internal server error" on other disk-related errors

Previously both were 400 "Bad request"

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Issue: https://github.com/restic/rest-server/issues/155

Checklist
---------

- [X] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [x] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/rest-server/blob/master/changelog/TEMPLATE))
- [X] I have run `gofmt` on the code in all commits
- [X] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/rest-server/commits/master)
- [x] I'm done, this Pull Request is ready for review

### CODE REVIEW

File: repo/repo.go [hunk 1]
```diff
@@ -603,7 +605,12 @@ func (h *Handler) saveBlob(w http.ResponseWriter, r *http.Request) {
 		if h.opt.Debug {
 			log.Print(err)
 		}
-		httpDefaultError(w, http.StatusBadRequest)
+		var pathError *os.PathError
+		if errors.As(err, &pathError) && (pathError.Err == syscall.ENOSPC || pathError.Err == syscall.EDQUOT) {
+			httpDefaultError(w, http.StatusInsufficientStorage)
+		} else {
+			httpDefaultError(w, http.StatusInternalServerError)
```
Comment (MichaelEischer) [inline repo/repo.go:612]: Please use `h.internalServerError(w, err)` here.

**Comment by wojas** [comment]
Thanks! The patch looks good, but the tests need to be updated, because they check for a 400.

File: repo/repo.go (hunk 1 above)
Comment (Enrico204) [reply repo/repo.go]: Done!

**Comment by Enrico204** [comment]
To fix all tests, I had to:

1. Add a variable for "file content does not match hash" error so I can distinguish between internal errors and bad request (to send a content that doesn't match the hash is bad request). Relevant lines:
   * https://github.com/restic/rest-server/pull/160/files#diff-65e557a13cfc70927d6259fdf0113fff370ecddebb867153d67ef9d0877d8db4R88-R91
   * https://github.com/restic/rest-server/pull/160/files#diff-65e557a13cfc70927d6259fdf0113fff370ecddebb867153d67ef9d0877d8db4R601
   * https://github.com/restic/rest-server/pull/160/files#diff-65e557a13cfc70927d6259fdf0113fff370ecddebb867153d67ef9d0877d8db4R612-R628
2. Change the error used in `TestAbortedRequest` function: now the function is using a "plausible" and declared error. The reason behind this change is because the `h.internalServerError` was crashing the server (due to `panic()` and `PanicOnError == true`) on stream reading error. Now the code detects whether the error is a network-related error (like unexpected EOF) and it use the correct HTTP status (400 Bad Request). Relevant change:
   * https://github.com/restic/rest-server/pull/160/files#diff-f3352cbd560f9467e98af5ecd129c7d1f43917302653fe5ceafbd8bc857f83c7R399
   * https://github.com/restic/rest-server/pull/160/files#diff-65e557a13cfc70927d6259fdf0113fff370ecddebb867153d67ef9d0877d8db4R612-R628

Tests should be OK. I also added some comments to clarify each `if` branch :-)

**Review by MichaelEischer — APPROVED** [review]
LGTM. Thanks for the test fixes!

**Pull request merged**